janitor/timeseries.py

```diff
@@ -0,0 +1,135 @@
+"""
+Time series-specific data testing and cleaning functions.
+"""
+
+import pandas as pd
+import pandas_flavor as pf
+from pandas.api.types import is_datetime64_any_dtype as is_datetime
+from collections import namedtuple
+from janitor import check
+
+
+def _flag_missing_timestamps(
+    df: pd.DataFrame,
+    frequency: str,
+    column_name: str,
+    first_time_stamp: pd.Timestamp,
+    last_time_stamp: pd.Timestamp,
+) -> namedtuple:
+    """
+    Test if timestamps are missing
+
+    Utility function to test if input data frame
+    is missing timestamps relative to expected timestamps.
+    They are generated based on the first_time_stamp,
+    last_time_stamp
+    and frequency.
+
+    :param df: data frame to test for missing timestamps
+    :param frequency: frequency i.e. sampling frequency
+        Acceptable frequency strings are available
+        `Reference <https://pandas.pydata.org/pandas-docs/stable/>`_
+        Check offset aliases under time series in user guide
+    :param column_name: column which has time series if not the index
+    :param first_time_stamp: timestamp expected to start from
+    :param last_time_stamp: timestamp expected to end with
+    :return: namedtuple with 3 attributes namely flag, raw_data and new_index
+        1. flag - boolean set to True if there are missing timestamps,
+            else set to False
+        2. raw_data - input data frame as is without any modifications
+        3. new_index - pd.DateTimeIndex that can be used to set the new index.
+            Defaults to None.
+            Assigned a value only when flag is set to True
+    """
+    # Declare a named tuple to hold results
+    MissingTimeStampFlag = namedtuple(
+        "MissingTimeStampFlag", ["flag", "raw_data", "new_index"]
+    )
+    result = {"flag": None, "raw_data": df.copy(deep=True), "new_index": None}
+
+    # Generate expected timestamps
+    expected_timestamps = pd.date_range(
+        start=first_time_stamp, end=last_time_stamp, freq=frequency
+    )
+
+    # Get actual timestamps
+    if column_name:
+        df = df.set_index(column_name)
+
+    df = df.sort_index()
+    actual_timestamps = df.index.array
+
+    # Check if they are the same
+    comparison_index = expected_timestamps.difference(actual_timestamps)
+    if comparison_index.empty:
+        result["flag"] = False
+    result["flag"] = True
```

Comment:
@UGuntupalli I found a chunk of time to go through the code in more detail, and I found this to be a bug.
It should have been the other way around:

```python
result["flag"] = True
if comparison_index.empty:
    result["flag"] = False
```

This bug resulted in `result["flag"]` never being false.

I took a deeper look at your functions' implementations, and I am going to propose a rework of the function. `_flag_missing_timestamps` seems to be doing one too many things here, and the `namedtuple` data structure, I think, might be unnecessary.

No worries, this is a fun puzzle to try to untangle. My goal here is to get functions to do one thing and _only_ one thing; I'll be eliminating the helper namedtuple, as it seems to be holding state where not necessary.